aws/sdk-codegen/src/test/kotlin/software/amazon/smithy/rustsdk/CredentialCacheConfigTest.kt

```diff
@@ -0,0 +1,150 @@
+        @endpointRuleSet({
+            "version": "1.0",
+            "rules": [{
+                          "type": "endpoint",
+                          "conditions": [{"fn": "isSet", "argv": [{"ref": "Region"}]}],
+                          "endpoint": { "url": "https://example.com" }
+                      }],
+            "parameters": {
+                "Region": { "required": false, "type": "String", "builtIn": "AWS::Region" },
+            }
+        })
+        service HelloService {
+            operations: [SayHello],
+            version: "1"
+        }
+
+        @optionalAuth
+        operation SayHello { input: TestInput }
+        structure TestInput {
+           foo: String,
+        }
+    """.asSmithyModel()
+
+    @Test
+    fun `config override for credentials`() {
+        awsSdkIntegrationTest(model, defaultToOrchestrator = true) { clientCodegenContext, rustCrate ->
+            val runtimeConfig = clientCodegenContext.runtimeConfig
+            val codegenScope = arrayOf(
+                *RuntimeType.preludeScope,
+                "Credentials" to AwsRuntimeType.awsCredentialTypesTestUtil(runtimeConfig)
+                    .resolve("Credentials"),
+                "CredentialsCache" to AwsRuntimeType.awsCredentialTypes(runtimeConfig).resolve("cache::CredentialsCache"),
+                "RuntimePlugin" to RuntimeType.smithyRuntimeApi(runtimeConfig).resolve("client::runtime_plugin::RuntimePlugin"),
+                "SharedCredentialsCache" to AwsRuntimeType.awsCredentialTypes(runtimeConfig).resolve("cache::SharedCredentialsCache"),
+            )
+            rustCrate.withModule(ClientRustModule.config) {
+                unitTest("test_overriding_credentials_provider_leads_to_shared_credentials_cache_in_layer") {
+                    rustTemplate(
+                        """
+                        use #{RuntimePlugin};
+
+                        let client_config = Config::builder().build();
+                        let config_override =
+                            Config::builder().credentials_provider(#{Credentials}::for_tests());
+                        let sut = ConfigOverrideRuntimePlugin {
+                            client_config: client_config.config().unwrap(),
+                            config_override,
+                        };
+                        let sut_layer = sut.config().unwrap();
+                        assert!(sut_layer
+                            .load::<#{SharedCredentialsCache}>()
+                            .is_some());
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                unitTest("test_overriding_credentials_cache_leads_to_shared_credentials_cache_in_layer") {
+                    rustTemplate(
+                        """
+                        use #{RuntimePlugin};
+
+                        // TODO(enableNewSmithyRuntimeCleanup): Should not need to provide a default once smithy-rs##2770
```

Comment:
What default is this referring to?

Reply:
Admittedly, the wording could be better. What default refers to is the client-level config in this case. Since `config_override` layer is overriding the client-level config, it sometimes needs to grab a value from the client-level config rather than a "pure" default value when newly constructing an orchestrator component within the `config_override` layer.

For instance, if a user just `config_override`s `credentials_cache`, we probably should respect whatever `credentials_provider` specified in the client-level config, rather than an arbitrarily chosen default (like `Credentials::for_tests`, although that's a bad example but hopefully you get the idea).

Reply:
Safest option is probably to require both of them to be set if one of them is set (for now).

Reply:
That works, updated in a78489d.